// mindee/lib/api/invoice.js
"use strict";

function asyncGeneratorStep(gen, resolve, reject, _next, _throw, key, arg) { try { var info = gen[key](arg); var value = info.value; } catch (error) { reject(error); return; } if (info.done) { resolve(value); } else { Promise.resolve(value).then(_next, _throw); } }

function _asyncToGenerator(fn) { return function () { var self = this, args = arguments; return new Promise(function (resolve, reject) { var gen = fn.apply(self, args); function _next(value) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "next", value); } function _throw(err) { asyncGeneratorStep(gen, resolve, reject, _next, _throw, "throw", err); } _next(undefined); }); }; }

var APIObject = require("./object");

var Input = require("../inputs");

class APIInvoice extends APIObject {
  constructor() {
    var apiToken = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : undefined;
    super(apiToken, "invoice");
    this.baseUrl = "".concat(this.baseUrl, "/invoices/");
  }
  /**
   * @param {Boolean} includeWords: , extract all words into http_response
   * @param {Boolean} cutPdf: Automatically reconstruct pdf with more than 4 pages
   * @param {String} inputType: String in {'path', 'stream', 'base64'}
   * @param {String} file: Receipt filepath (allowed jpg, png, tiff, pdf)
   * @param {String} version: expense_receipt api version
   * @returns {Response} Wrapped response with Receipts objects parsed
   */


  parse(_ref) {
    var _superprop_getParse = () => super.parse,
        _superprop_get_request = () => super._request,
        _this = this;

    return _asyncToGenerator(function* () {
      var {
        input,
        inputType = "path",
        filename = undefined,
        version = "2",
        cutPdf = true,
        includeWords = false
      } = _ref;

      _superprop_getParse().call(_this);

      var inputFile = new Input({
        file: input,
        inputType,
        filename,
        cutPdf
      });
      yield inputFile.init();
      var url = "v".concat(version, "/predict");
      return yield _superprop_get_request().call(_this, url, inputFile, includeWords);
    })();
  }
  /** 
    @param {String} inputFile - Input object
    @param {} response - HTTP response
    @param {Document} documentType - Document class in {"Receipt", "Invoice", "Financial_document"}
    @returns {Response}
  */


  wrapResponse(inputFile, response, documentType) {
    var _result$httpResponse$, _result$httpResponse$2, _result$httpResponse$3, _result$httpResponse$4, _result$httpResponse$5;

    var result = super.wrapResponse(inputFile, response, documentType);
    result.documentType = ((_result$httpResponse$ = result.httpResponse.data) === null || _result$httpResponse$ === void 0 ? void 0 : (_result$httpResponse$2 = _result$httpResponse$.predictions) === null || _result$httpResponse$2 === void 0 ? void 0 : (_result$httpResponse$3 = _result$httpResponse$2[0]) === null || _result$httpResponse$3 === void 0 ? void 0 : (_result$httpResponse$4 = _result$httpResponse$3.document_type) === null || _result$httpResponse$4 === void 0 ? void 0 : (_result$httpResponse$5 = _result$httpResponse$4.value) === null || _result$httpResponse$5 === void 0 ? void 0 : _result$httpResponse$5.toLowerCase()) || result.documentType;
    return result;
  }

}

module.exports = APIInvoice;
